src/types.ts

```diff
@@ -1051,21 +1051,23 @@ export class Signature {
   isAssignableTo(target: Signature, checkCompatibleOverride: bool = false): bool {
     let thisThisType = this.thisType;
     let targetThisType = target.thisType;
-    if (checkCompatibleOverride) {
-      // check kind of `this` type
-      if (thisThisType) {
-        if (!targetThisType || !thisThisType.canExtendOrImplement(targetThisType)) {
+
+    if (
+      (thisThisType == null && targetThisType != null) ||
+      (thisThisType != null && targetThisType == null)
+    ) {
+      return false;
+    }else if(thisThisType != null && targetThisType != null){
+      if(checkCompatibleOverride){
```

Comment:
@dcodeIO @MaxGraey Dear Maintainers, we do you prefer. The existing check has duplicate code such as
```typescript
} else if (targetThisType) {
        return false;
}
```
It's not a key point, I will follow the style which maintainer like.

Reply:
@CountBleck could you also review this style? In follow code, everything in `if` and `else` are same except `thisThisType.canExtendOrImplement(targetThisType)` and `!argetThisType.isAssignableTo(thisThisType)`. I'm afraid duplicate code doesn't look like better format. But since "better format" and "more readable"  are a bit subjective, I will follow the community's favourite. 

```typescript
    if (checkCompatibleOverride) {
      // check kind of `this` type
      if (thisThisType) {
        if (!targetThisType || !thisThisType.canExtendOrImplement(targetThisType)) {
          return false;
        }
      } else if (targetThisType) {
        return false;
      }
    } else {
      // check `this` type (invariant)
      if (thisThisType) {
        if (!targetThisType || !targetThisType.isAssignableTo(thisThisType)) {
          return false;
        }
      } else if (targetThisType) {
        return false;
      }
    }
```

Reply:
Hm, I see your point. It's that versus...
```ts
    if (thisThisType) {
      if (!targetThisType) return false;
      if (checkCompatibleOverride) {
        // check kind of `this` type
        if (!thisThisType.canExtendOrImplement(targetThisType)) return false;
      } else {
        // check `this` type (contravariant)
        if (!targetThisType.isAssignableTo(thisThisType)) return false;
      }
    } else if (targetThisType) {
      return false;
    }
```
...which looks a bit weird.

This also looks a bit weird:
```ts
    if (thisThisType) {
      if (!targetThisType) return false;
      let compatibleThisType = checkCompatibleOverride
        ? thisThisType.canExtendOrImplement(targetThisType))
        : targetThisType.isAssignableTo(thisThisType);
      if (!compatibleThisType) return false;
    } else if (targetThisType) {
      return false;
    }
```

Reply:
Both `checkCompatibleOverride==true` and `checkCompatibleOverride==false` cases needs to return false when one of `targetThisType` and `thisThisType` is null. It's the reason why I put this check in advanced. Otherwise, we need a `if (!targetThisType) return false;` in if branch and another `else if (targetThisType) return false.`

But I think your ` let compatibleThisType = checkCompatibleOverridethisThisType.canExtendOrImplement(targetThisType)): targetThisType.isAssignableTo(thisThisType);` is very good style. We can use it.

Does this version looks better?
```typescript
if (
      (thisThisType == null && targetThisType != null) ||
      (thisThisType != null && targetThisType == null)
    ) {
      return false;
    }else if(thisThisType != null && targetThisType != null){

      const compatibleThisType = checkCompatibleOverride ? thisThisType.canExtendOrImplement(targetThisType)
        : targetThisType.isAssignableTo(thisThisType);
      if (!compatibleThisType) {
        return false;
      }
    }
```

Reply:
> Both `checkCompatibleOverride==true` and `checkCompatibleOverride==false` cases needs to return false when one of `targetThisType` and `thisThisType` is null.

Look closely at the code. That check is already there. Your version of the check is just more verbose.

> Does this version looks better?
I think you need to get introduced to some sort of formatter, because your whitespace is all over the place :P
That alone makes that version look worse.

Reply:
I guess your concern is majorly about the `(thisThisType == null && targetThisType != null) || (thisThisType != null && targetThisType == null)`. It's the "verbose" you said, because it checks `thisThisType == null` and `thisThisType != null` twice. But `if (!targetThisType) return false` and `else if (targetThisType) return false` also both `return false`, it's perhaps the "wired" you mean.

Then how abut this?
```typescript
if (
      (thisThisType || targetThisType) && !(thisThisType && thisThisType)
    )
```

Reply:
sure，we can use this one. Currently I'm away of my laptop. I will commit fix of all your comments tomorrow